zetta_utils/augmentations/misalign.py

```diff
@@ -0,0 +1,165 @@
+    ) -> VolumetricIndex:
+        if mode != "read":
+            raise NotImplementedError()
+        disp_x_in_unit_magn = random.uniform(self.disp_min_in_unit, self.disp_max_in_unit)
+        disp_y_in_unit_magn = random.uniform(self.disp_min_in_unit, self.disp_max_in_unit)
+        if self.disp_in_unit_must_be_divisible_by is not None:
+            disp_x_in_unit_magn = math.floor(
+                disp_x_in_unit_magn / self.disp_in_unit_must_be_divisible_by
+            )
+            disp_y_in_unit_magn = math.floor(
+                disp_y_in_unit_magn / self.disp_in_unit_must_be_divisible_by
+            )
+        disp_x_in_unit = disp_x_in_unit_magn * random.choice([1, -1])
+        disp_y_in_unit = disp_y_in_unit_magn * random.choice([1, -1])
+
+        disp_x_in_idx_res = disp_x_in_unit / idx.resolution[0]
+        disp_y_in_idx_res = disp_y_in_unit / idx.resolution[1]
+
+        # self.prepared_disp = (disp_x, disp_y)
+
+        start_offset = [0, 0, 0]
+        end_offset = [0, 0, 0]
+        if disp_x_in_idx_res > 0:
+            start_offset[0] = -disp_x_in_idx_res
+        else:
+            end_offset[0] = -disp_x_in_idx_res
+        if disp_y_in_idx_res > 0:
+            start_offset[1] = -disp_y_in_idx_res
+        else:
+            end_offset[1] = -disp_y_in_idx_res
+
+        idx = idx.translated_start(Vec3D[int](*start_offset)).translated_end(
+            Vec3D[int](*end_offset)
+        )
+        self.prepared_disp_fraction = (
+            disp_x_in_idx_res / idx.shape[0],
+            disp_y_in_idx_res / idx.shape[1],
+        )
+        return idx
+
+    def process_data(self, data: T, mode: Literal["read", "write"]) -> T:
+        if mode != "read":
+            raise NotImplementedError()
+
+        if isinstance(data, torch.Tensor):
+            z_size = data.shape[-1]
+        else:
+            if not self.keys_to_apply:
+                raise ValueError(
+                    "`keys_to_apply` must be a non-empty list of strings when "
+                    "applying to data of type `dict`"
+                )
+            sizes = [data[k].shape for k in self.keys_to_apply]
+            if self.disp_in_unit_must_be_divisible_by is None and not all(
+                e[0:2] == sizes[0][0:2] for e in sizes
+            ):
+                raise ValueError(
+                    "When `disp_in_unit_must_be_divisible_by` is not specified, "
+                    "XY sizes of all keys affected by misalignmnet augmentation "
+                    f"must be identical. Got: {sizes}"
+                )
+            z_size = max(size[-1] for size in sizes)
+
+        z_chosen = random.randint(0, z_size - 1)
+
+        def _process_tensor(tensor: torch.Tensor) -> torch.Tensor:
+            if tensor.size(-1) < z_size:
+                z_diff = z_size - tensor.size(-1)
+                assert z_diff > 0
+                assert z_diff % 2 == 0
+                this_z_chosen = z_chosen - (z_diff // 2)
```

Comment:
The logic here is broken. If `z_chosen` is 0, then `this_z_chosen` becomes negative. Creating a `slice` with a negative `z_chosen` is a logical error.